### PR TITLE
Octave and Sundials27

### PR DESCRIPTION
**Pull request by schoeps**
Octave with the most recent patches for macOS. In particular "--HEAD" is now working pretty neatly on macOS. Sundials (in version 2.7) is needed for head.

### CODE REVIEW

**Comment by dpo** [comment]
There we go.

**Comment by dpo** [comment]
There are some audits:
```
Error: 5 problems in 1 formula
dpo/openblas/octave:
  * C: 48: col 3: `option` (line 48) should be put before `depends_on` (line 18)
  * W: 55: col 10: Odd `else` layout detected. Did you mean to use `elsif`?
  * C: 58: col 7: Inconsistent indentation detected.
  * C: 64: col 7: Inconsistent indentation detected.
  * C: 83: col 38: Trailing whitespace detected.
```
and it appears X11 isn't installed. I need to investigate that.

File: sundials27.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  depends_on "cmake" => :build
+  depends_on "gcc" # for gfortran
+  depends_on "open-mpi" if build.with? "mpi"
+  depends_on "dpo/openblas/suite-sparse"
+  depends_on "openblas"
+
+  fails_with :clang if build.with? "openmp"
+
+  def install
+    blas = "-L#{Formula["openblas"].opt_lib}"
```
Comment (dpo) [inline sundials27.rb]: Is there a `-lopenblas` missing here? Currently, CMake says
```
-- Checking if Lapack works... FAILED
WARNING: LAPACK not functional.
```

**Comment by dpo** [comment]
Does Sundials require Python?
```
-- Could NOT find PythonInterp (missing: PYTHON_EXECUTABLE) 
CMake Error at config/SundialsAddTest.cmake:27 (IF):
  if given arguments:

    "LESS" "3"

  Unknown arguments specified
Call Stack (most recent call first):
  CMakeLists.txt:941 (INCLUDE)
```

**Comment by dpo** [comment]
And if you wouldn't mind rebasing yet again.

**Comment by schoeps** [comment]
sundials does not need python to work. It's just for testing. However, I would think we shuld run tests in the future for scientific packages, so I added python as a build dependency.

(Do you know how to do a python3 dependency properly with cmake? I looked at vtk.rb and got scared)

**Comment by dpo** [comment]
Looking good. Is `-lm` missing in the sundials test? Not sure why it fails on linux but not on macOS:
```
==> brew test dpo/openblas/sundials27 --verbose
Testing dpo/openblas/sundials27
==> /home/linuxbrew/.linuxbrew/bin/gcc-5 -I/home/linuxbrew/.linuxbrew/Cellar/sundials27/2.7.0/include test_nvector.c sundials_nvector.c test_nvector_serial.c -L/home/linuxbrew/.linuxbrew/Cellar/sundials27/2.7.0/lib -lsundials_nvecserial
.
Last 150 lines from /home/linuxbrew/.cache/Homebrew/Logs/sundials27/test.01.gcc-5:
2018-01-28 11:48:52 +0000

/home/linuxbrew/.linuxbrew/bin/gcc-5
-I/home/linuxbrew/.linuxbrew/Cellar/sundials27/2.7.0/include
test_nvector.c
sundials_nvector.c
test_nvector_serial.c
-L/home/linuxbrew/.linuxbrew/Cellar/sundials27/2.7.0/lib
-lsundials_nvecserial

/home/linuxbrew/.linuxbrew/Cellar/sundials27/2.7.0/lib/libsundials_nvecserial.so: undefined reference to `exp'
/home/linuxbrew/.linuxbrew/Cellar/sundials27/2.7.0/lib/libsundials_nvecserial.so: undefined reference to `pow'
/home/linuxbrew/.linuxbrew/Cellar/sundials27/2.7.0/lib/libsundials_nvecserial.so: undefined reference to `sqrt'
collect2: error: ld returned 1 exit status
```

I'm not sure about CMake. The part of `vtk.rb` that concerns python3 is quite short though.

**Comment by schoeps** [comment]
Ok, I added "-lm" to the test to see if Linux needs it.

regarding python: in [`vtk.rb`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/vtk.rb) needs lines 61-71 to figure which python to use. Here, i do not care whether 2 or 3 but I recall that homebrew will change how to treat python.

**Comment by dpo** [comment]
That did it. It's all looking good now. Is this ready to be merged from your end?

I'm not sure about python.

**Comment by schoeps** [comment]
yes, I am happy. I will now start playing with `getdp` and `gmsh` (+deps, e.g. `petsc`) and eventually move them to this repository. I also like to take care on proper testing of all builds, e.g. run (the time consuming) official test suites automatically and make the build fail if issues pop up. Currently, we check only for basic errors like wrong linking. We should aim for more :)

**Comment by dpo** [comment]
I agree. Many thanks for all the time and hard work!

**Comment by apjanke** [comment]
Hi folks. Do you know how much demand there is in the Octave user community for the solvers that depend on SUNDIALS? (`ode5i` and `ode5s`, I think.) I'm looking to fix the Octave SUNDIALS dependency in Homebrew core over at https://github.com/Homebrew/homebrew-core/pull/29387 and there's a question of how wanted it is in the first place.

**Comment by schoeps** [comment]
I would say that sundials is important. ode15s and ode15i are the only Marlab-compatible stiff ODE solvers in Octave. There has been an effort to move to the newer version (http://savannah.gnu.org/bugs/?52475) but it's obviously not stable yet.

**Pull request merged**